### PR TITLE
Feature/native date picker

### PR DESCRIPTION
**Pull request by mario-subo**
Native datepicker

### CODE REVIEW

**Comment by mario-subo** [comment]
Originally this PR was supposed to be merged into develop but now the base branch is the facelift so it includes some other changes which can be ignored until the facelift  (new base branch) is aligned with develop.

File: src/hooks/useDynamicForm.tsx [hunk 1]
```diff
@@ -168,20 +169,26 @@ export function useDynamicForm(sections: DynamicFormSection[]): [() => JSX.Eleme
...
-                            month: targetDate.getMonth() + 1,
-                            day: targetDate.getDate(),
-                        } as DynamicFormDate;
-                    }
+                case "Checkbox": {
+                    const targetValue: boolean = (e as React.ChangeEvent<HTMLInputElement>).target.checked;
+                    newValue = targetValue;
                     break;
                 }
+                // case "Checkbox":
```
Comment (DanSallau) [inline src/hooks/useDynamicForm.tsx:177]: Is there any reason why this is commented ? Can we remove it entirely ?

File: src/pages/docs/datepicker.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                    rulerKey: "forceCustom",
+                    condition: true,
+                    label: "Locale code:",
+                    description: "Set a locale for the custom picker (defaults to system locale).",
+                    controlType: "Text",
+                },
+            ],
+        },
+    ]);
+    const code: string = `<Datepicker value={myDateValue} onChange={(newDate: Date) => setMyDateValue(newDate)} />`;
```
Comment (DanSallau) [inline src/pages/docs/datepicker.tsx:37]:
```suggestion
    const code: string = React.useMemo(() => `<Datepicker value={myDateValue} onChange={(newDate: Date) => setMyDateValue(newDate)} />`, []);
```

File: lib/src/Datepicker/Datepicker.test.tsx [hunk 3]
```diff
@@ -3,84 +3,72 @@ import { shallow, mount, ShallowWrapper, ReactWrapper } from "enzyme";
...
-    it("Should render with custom locale", () => {
-        const locale: string = "sv-se";
-        const value: Date = new Date();
-        const [year, month, day] = [2015, 11, 25];
-        value.setFullYear(year, month, day);
-        mountedWrapper.setProps({ locale, value });
-        expect(mountedWrapper.find(`input[name="${props.name}"]`).getElement().props.value).toEqual(`${year}-${month + 1}-${day}`);
-    });
+    // const props: DatepickerProps = {
+    //     name: "myDatepicker",
```
Comment (DanSallau) [inline lib/src/Datepicker/Datepicker.test.tsx:7]: Don't forget to enable test.

**Review by DanSallau — APPROVED** [review]

File: lib/src/Datepicker/Datepicker.test.tsx [hunk 4]
```diff
@@ -4,83 +4,70 @@ import { Datepicker, DatepickerProps } from ".";
...
+    //     expect(wrapper.find(".date-placeholder").length).toBeTruthy();
+    // });
+    // it("Should render with custom locale", () => {
+    //     const locale: string = "sv-se";
+    //     const value: Date = new Date();
+    //     const [year, month, day] = [2015, 11, 25];
+    //     value.setFullYear(year, month, day);
+    //     mountedWrapper.setProps({ locale, value });
+    //     expect(mountedWrapper.find(`input[name="${props.name}"]`).getElement().props.value).toEqual(`${year}-${month + 1}-${day}`);
+    // });
```
Comment (yousifalraheem) [inline lib/src/Datepicker/Datepicker.test.tsx:72]: Maybe you can finish unit test before merging?

**Review by yousifalraheem — CHANGES_REQUESTED** [review]

File: lib/src/Datepicker/Datepicker.test.tsx (hunk 4 above)
Comment (mario-subo) [reply lib/src/Datepicker/Datepicker.test.tsx:72]: All the unit tests are failing because of some typescript config issue I think, need to investigate.

Comment (mario-subo) [reply lib/src/Datepicker/Datepicker.test.tsx:72]: ok fixed in base branch

File: lib/src/Datepicker/formatters.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+export function padNumber(value: number, isYear: boolean = false): string {
+    if (isNumber(value)) {
+        return isYear ? `0000${value}`.substr(-4, 4) : `00${value}`.substr(-2, 2);
+    } else {
+        return "";
+    }
+}
+
+export function toInteger(value: any): number {
+    return parseInt(`${value}`, 10);
```
[the remark this answers is not in the document]
Comment (mario-subo) [reply lib/src/Datepicker/formatters.ts:10]: yeah not sure why that ended up like that

File: lib/src/Datepicker/formatters.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+export function isNumber(value: any): value is number {
+    return !isNaN(toInteger(value));
+}
+
+export function areNumbers(values: any[]): boolean {
+    return values.every((val) => isNumber(val));
+}
```
[the remark this answers is not in the document]
Comment (mario-subo) [reply lib/src/Datepicker/formatters.ts]: Great suggestion!

[the remark this answers is not in the document]
Comment (mario-subo) [reply lib/src/Datepicker/formatters.ts]: Should be like this tho :P 
```
export function isNumber(...values): boolean {
    return values.every((val) => !isNaN(toInteger(val)));
}
```

**Review by kherP — APPROVED** [review]

File: lib/src/Datepicker/formatters.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+export function padNumber(value: number, isYear: boolean = false): string {
+    if (isNumber(value)) {
+        return isYear ? `0000${value}`.substr(-4, 4) : `00${value}`.substr(-2, 2);
+    } else {
+        return "";
+    }
+}
```
Comment (yousifalraheem) [inline lib/src/Datepicker/formatters.ts:7]: Use this syntax, much cleaner
```javascript
(num).toLocaleString('en', { minimumIntegerDigits: 2, useGrouping: false })
```

File: lib/src/Datepicker/Datepicker.tsx [hunk 8]
```diff
@@ -1,65 +1,334 @@
...
+    forceCustom?: boolean;
+    /** Property sets whether datepicker is disabled */
+    disabled?: boolean;
+    /** Minimum range of date that can be selected */
+    min?: Date;
+    /** Maximum range of date that can be selected */
+    max?: Date;
+    /** Locale of datepicker */
+    localeCode?: string;
+}
```
[the remark this answers is not in the document]
Comment (mario-subo) [reply lib/src/Datepicker/Datepicker.tsx:26]: How do you suggest we cater for the fact that this component includes a fallback custom datepicker (with combination of <input /> and <select /> wrapped in a <div />)?

File: lib/src/Datepicker/Datepicker.tsx [hunk 9]
```diff
@@ -1,65 +1,334 @@
...
+                onChange={handleOnChange}
+            />
+        );
+    } else {
+        return <>{renderCustomDatepicker(value, monthPicker, customPickerOrder(), unitNames, disabled, monthNames())}</>;
+    }
+
+    // return (
+    //     <>fallback</>
+    // )
```
[the remark this answers is not in the document]
Comment (mario-subo) [reply lib/src/Datepicker/Datepicker.tsx:329]: removed

File: lib/src/Datepicker/formatters.ts (hunk 7 above)
Comment (mario-subo) [reply lib/src/Datepicker/formatters.ts:7]: Nice suggestion, thanks! I thought it is not well supported by older browsers but it looks like it is.

File: lib/src/Datepicker/Datepicker.tsx [hunk 10]
```diff
@@ -1,65 +1,334 @@
...
+    };
+
+    const [customDay, setCustomDay] = React.useState<number>(initCustomDay(value, monthPicker));
+
+    const handleChangeCustomDay = (e: React.ChangeEvent<HTMLInputElement>): void => {
+        const v: number = e.target?.value ? Number(e.target.value) : null;
+        if (!monthPicker) {
+            setCustomDay(v);
+        }
+    };
```
[the remark this answers is not in the document]
Comment (mario-subo) [reply lib/src/Datepicker/Datepicker.tsx:85]: good point

**Review by yousifalraheem — APPROVED** [review]

File: lib/src/Datepicker/Datepicker.tsx [hunk 11]
```diff
@@ -1,65 +1,364 @@
...
-    placeholder?: string;
-    showLeadingZeros?: boolean;
+interface OverriddenNativeProps extends React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement> {
+    value: any;
+    min?: any;
+    max?: any;
+    onChange: any;
+}
+
+export interface DatepickerProps extends OverriddenNativeProps {
```
Comment (yousifalraheem) [inline lib/src/Datepicker/Datepicker.tsx:13]: Can you use the same syntax we've been using with the rest of our refactored components?

```typescript
export type DatepickerProps = JSX.IntrinsicElements["input"] & {
```

**Review by yousifalraheem — CHANGES_REQUESTED** [review]

**Review by yousifalraheem — APPROVED** [review]

**Pull request merged**